File: lib/EncodedVideoChunk.ts
/**
 * EncodedVideoChunk - TypeScript wrapper for native EncodedVideoChunk
 * @see spec/context/EncodedVideoChunk.md
 */

import type { AllowSharedBufferSource, EncodedVideoChunkInit, EncodedVideoChunkType } from '../types/webcodecs';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const bindings = require('bindings')('webcodecs');

export class EncodedVideoChunk {
  private readonly _native: unknown;

  constructor(init: EncodedVideoChunkInit) {
    this._native = new bindings.EncodedVideoChunk(init);
  }

  get type(): EncodedVideoChunkType {
    return (this._native as Record<string, unknown>).type as EncodedVideoChunkType;
  }

  get timestamp(): number {
    return (this._native as Record<string, unknown>).timestamp as number;
  }

  get duration(): number | null {
    return (this._native as Record<string, unknown>).duration as number | null;
  }

  get byteLength(): number {
    return (this._native as Record<string, unknown>).byteLength as number;
  }


  copyTo(destination: AllowSharedBufferSource): void {
    return (this._native as Record<string, Function>).copyTo(destination) as void;
  }

}

File: src/encoded_video_chunk.cpp
#include "encoded_video_chunk.h"

namespace webcodecs {

Napi::FunctionReference EncodedVideoChunk::constructor;

Napi::Object EncodedVideoChunk::Init(Napi::Env env, Napi::Object exports) {
  Napi::Function func = DefineClass(env, "EncodedVideoChunk",
                                    {
                                        InstanceAccessor<&EncodedVideoChunk::GetType>("type"),
                                        InstanceAccessor<&EncodedVideoChunk::GetTimestamp>("timestamp"),
                                        InstanceAccessor<&EncodedVideoChunk::GetDuration>("duration"),
                                        InstanceAccessor<&EncodedVideoChunk::GetByteLength>("byteLength"),
                                        InstanceMethod<&EncodedVideoChunk::CopyTo>("copyTo"),
                                    });

  constructor = Napi::Persistent(func);
  constructor.SuppressDestruct();
  exports.Set("EncodedVideoChunk", func);
  return exports;
}

EncodedVideoChunk::EncodedVideoChunk(const Napi::CallbackInfo& info) : Napi::ObjectWrap<EncodedVideoChunk>(info) {
  Napi::Env env = info.Env();

  // [SPEC] Constructor Algorithm
  /*
   * Initialize internal slots.
   */

  // TODO(impl): Implement Constructor & Resource Allocation
}

EncodedVideoChunk::~EncodedVideoChunk() { Release(); }

void EncodedVideoChunk::Release() {
  // Release packet (RAII handles av_packet_unref)
  packet_.reset();
}

// --- Attributes ---

Napi::Value EncodedVideoChunk::GetType(const Napi::CallbackInfo& info) {
  // TODO(impl): Return type
  return info.Env().Null();
}

Napi::Value EncodedVideoChunk::GetTimestamp(const Napi::CallbackInfo& info) {
  // TODO(impl): Return timestamp
  return info.Env().Null();
}

Napi::Value EncodedVideoChunk::GetDuration(const Napi::CallbackInfo& info) {
  // TODO(impl): Return duration
  return info.Env().Null();
}

Napi::Value EncodedVideoChunk::GetByteLength(const Napi::CallbackInfo& info) {
  // TODO(impl): Return byteLength
  return info.Env().Null();
}

// --- Methods ---

Napi::Value EncodedVideoChunk::CopyTo(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  // [SPEC] Algorithm
  /*
   * See spec/context file.
   */

  // TODO(impl): Implement method logic
  return env.Undefined();
}

}  // namespace webcodecs
